tests/unit/websocket.cpp

```diff
@@ -0,0 +1,262 @@
+    const auto response = client.Read();
+
+    EXPECT_TRUE(response == auth_success) << "Response was: " << response;
+  }
+  {
+    auto client = Client();
+    EXPECT_NO_THROW(client.Connect(host, port));
+    EXPECT_NO_THROW(client.Write(R"({"username": "user" "password": "123"})"));
+    const auto response1 = nlohmann::json::parse(client.Read());
+    const auto message_header1 = response1[kResponseMessage].get<std::string>();
+    const auto message_first_part1 = message_header1.substr(0, message_header1.find(": "));
+
+    EXPECT_FALSE(response1[kResponseSuccess]);
+    EXPECT_EQ(message_first_part1, auth_fail) << "Message was: " << message_header1;
+
+    EXPECT_NO_THROW(client.Connect(host, port));
+    EXPECT_NO_THROW(client.Write(R"({"username": "user", "password": "123"})"));
+    const auto response2 = client.Read();
+    EXPECT_TRUE(response2 == auth_success) << "Response was: " << response2;
+  }
+  {
+    auto client1 = Client();
+    auto client2 = Client();
+
+    EXPECT_NO_THROW(client1.Connect(host, port));
+    EXPECT_NO_THROW(client2.Connect(host, port));
+
+    EXPECT_NO_THROW(client1.Write(R"({"username": "user" "password": "123"})"));
+    EXPECT_NO_THROW(client2.Write(R"({"username": "user", "password": "123"})"));
+
+    const auto response1 = nlohmann::json::parse(client1.Read());
+    const auto message_header1 = response1[kResponseMessage].get<std::string>();
+    const auto message_first_part1 = message_header1.substr(0, message_header1.find(": "));
+    EXPECT_FALSE(response1[kResponseSuccess]);
+    EXPECT_EQ(message_first_part1, auth_fail) << "Message was: " << message_header1;
+
+    const auto response2 = client2.Read();
+    EXPECT_TRUE(response2 == auth_success) << "Response was: " << response2;
+  }
+
+  EXPECT_NO_THROW(websocket_server.Shutdown());
+  EXPECT_NO_THROW(websocket_server.AwaitShutdown());
+}
+
+TEST(WebSocketServer, WebsocketAuthenticationFails) {
+  MockAuth auth{false, true, true};
+  communication::websocket::Server websocket_server({"0.0.0.0", 0}, &auth);
+  const auto port = std::to_string(websocket_server.GetEndpoint().port());
+  constexpr auto host = "0.0.0.0";
+
+  constexpr auto auth_fail = R"({"message":"Authentication failed!","success":false})";
+  websocket_server.Start();
+  {
+    auto client = Client();
+    EXPECT_NO_THROW(client.Connect(host, port));
+    EXPECT_NO_THROW(client.Write(R"({"username": "user", "password": "123"})"));
+
+    const auto response = client.Read();
+    EXPECT_EQ(response, auth_fail);
+  }
+  EXPECT_NO_THROW(websocket_server.Shutdown());
+  EXPECT_NO_THROW(websocket_server.AwaitShutdown());
+}
+
+TEST(WebSocketServer, WebsocketAuthorizationFails) {
+  MockAuth auth{true, false, true};
+  communication::websocket::Server websocket_server({"0.0.0.0", 0}, &auth);
+  constexpr auto host = "0.0.0.0";
+  const auto port = std::to_string(websocket_server.GetEndpoint().port());
```

Comment:
I think this part and the destruction of the server could be extracted to a [test fixture](https://github.com/google/googletest/blob/main/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests). I think providing full access to all members of `MockAuth` doesn't harm anybody.

Using a test fixture might make it possible to put the test cases of `WebsocketAuthenticationWhenAuthPasses` into different tests.